core/keystore/src/lib.rs

```diff
@@ -14,185 +14,180 @@
+#[derive(Clone)]
+pub struct LocalStore<TPair: Pair> {
+	keys: Arc<Mutex<HashMap<TPair::Public, (TPair, Instant)>>>,
+}
 
-	/// Load a key file with given public key.
-	pub fn load<TPair: Pair>(&self, public: &TPair::Public, password: &str) -> Result<TPair> {
-		if let Some(pair) = self.get_pair(public)? {
-			return Ok(pair)
-		}
+impl<TPair> LocalStore<TPair>
+where
+	TPair: Pair + Send + Clone,
+	TPair::Public: Send + Clone,
+{
+	/// Creates a new `LocalStore`.
+	pub fn new() -> Self {
+		let session_store = Self {
+			keys: Default::default(),
+		};
+		session_store
+	}
 
-		let path = self.key_file_path::<TPair>(public);
-		let file = File::open(path)?;
+	/// Gets the key matching the public key.
+	pub fn get_key(&self, public: &TPair::Public) -> Option<TPair> {
+		let pair = {
+			let mut keys = self.keys.lock();
+			if let Some(value) = keys.get_mut(public) {
+				value.1 = Instant::now();
+				Some(value.0.clone())
+			} else {
+				None
+			}
+		};
+		self.clean();
+		pair
+	}
 
-		let phrase: String = ::serde_json::from_reader(&file)?;
-		let (pair, _) = TPair::from_phrase(&phrase, Some(password))
-			.ok().ok_or(Error::InvalidPhrase)?;
-		if &pair.public() != public {
-			return Err(Error::InvalidPassword);
-		}
-		Ok(pair)
+	/// Find keys matching a predicate.
+	pub fn get_keys(&self, f: impl Fn(&TPair::Public) -> bool) -> Vec<TPair> {
+		let filtered_keys = {
+			let mut keys = self.keys.lock();
+			let filtered_keys = keys.iter()
+				.filter(|(public, _)| f(public))
+				.map(|(_, (pair, _))| pair.clone())
+				.collect::<Vec<_>>();
+			for key in &filtered_keys {
+				keys.get_mut(&key.public())
+					.expect("We hold a lock; qed").1 = Instant::now();
+			}
+			filtered_keys
+		};
+		self.clean();
+		filtered_keys
 	}
 
-	/// Get public keys of all stored keys.
-	pub fn contents<TPublic: Public>(&self) -> Result<Vec<TPublic>> {
-		let mut public_keys: Vec<TPublic> = self.additional.keys()
-			.filter_map(|(ty, public)| {
-				if *ty != TPublic::KEY_TYPE {
-					return None
-				}
-				Some(TPublic::from_slice(public))
-			})
-			.collect();
-
-		let key_type: [u8; 4] = TPublic::KEY_TYPE.to_le_bytes();
-		for entry in fs::read_dir(&self.path)? {
-			let entry = entry?;
-			let path = entry.path();
-
-			// skip directories and non-unicode file names (hex is unicode)
-			if let Some(name) = path.file_name().and_then(|n| n.to_str()) {
-				match hex::decode(name) {
-					Ok(ref hex) => {
-						if hex[0..4] != key_type { continue	}
-						let public = TPublic::from_slice(&hex[4..]);
-						public_keys.push(public);
-					}
-					_ => continue,
-				}
+	/// Cleanup session store.
+	fn clean(&self) {
+		let mut keys = self.keys.lock();
+		if keys.len() > 1 {
+			let duration = Duration::new(INACTIVITY_PERIOD, 0);
+			let new_keys = keys.iter().filter(|(_, (_, instant))| {
+				Instant::now().duration_since(instant.clone()) < duration
+			}).map(|(k, v)| (k.clone(), v.clone())).collect::<HashMap<_, _>>();
+			if new_keys.len() > 0 {
```

Comment:
Why do we keep the old keys?

Reply:
to prevent a validator from being unstaked but later restaked. this could lead to the validator getting slashed if he no longer has his session key